cpp/src/fil/infer.cu

```diff
@@ -172,102 +206,190 @@ struct tree_aggregator_t {
     : tmp_storage(finalize_workspace) {}
 
   __device__ __forceinline__ void accumulate(
-    vec<NITEMS, float> single_tree_prediction, int tree) {
+    vec<NITEMS, float> single_tree_prediction, int tree, int num_rows) {
+    for (int i = num_rows; i < NITEMS; ++i) acc[i] = 0.0f;
     acc += single_tree_prediction;
   }
 
   __device__ __forceinline__ void finalize(float* out, int num_rows,
-                                           int output_stride) {
+                                           int output_stride,
+                                           output_t transform, int num_trees) {
     __syncthreads();
-    typedef typename BlockReduce<NITEMS>::TempStorage TempStorage;
-    acc = BlockReduce<NITEMS>(*(TempStorage*)tmp_storage).Sum(acc);
+    acc = block_reduce(acc, vectorized(cub::Sum()), tmp_storage);
     if (threadIdx.x > 0) return;
 #pragma unroll
-    for (int row = 0; row < num_rows; ++row)
-      out[row * output_stride] = acc[row];
+    for (int row = 0; row < NITEMS; ++row)
```

Comment:
Just for my understanding as I look over the NITEMS changes - where do we guarantee that NITEMS is always >= nrows?

Reply:
line 512 in infer.cu
https://github.com/rapidsai/cuml/pull/2894/files/282a4befcff5f198a4b5bce8a65038a58df2ea7c#diff-7e618cd37d0a3d4eac0025626bac28016d611855f09fa5df6e9c9f61439db100R512
we pass this `min((size_t)NITEMS, params.num_rows - block_row0)` as `num_rows` into `tree_aggregator_t`